LoopUp.Siesta/Patch/JsonPatchDocumentHelpers.cs

```diff
@@ -0,0 +1,193 @@
+    /// Helper class for making and working with JsonPatchDocuments.
+    /// </summary>
+    public static class JsonPatchDocumentHelpers
+    {
+        /// <summary>
+        /// Generate a JsonPatchDocument through the comparison of two versions of the same class.
+        /// </summary>
+        /// <param name="original">Original version of the class.</param>
+        /// <param name="modified">New version of the class.</param>
+        /// <typeparam name="T">The type of the class.</typeparam>
+        /// <returns>A JsonPatchDocument for the two objects.</returns>
+        public static JsonPatchDocument GeneratePatchDocument<T>(T original, T modified)
+            where T : class
+        {
+            var patchDocument = new JsonPatchDocument();
+
+            PopulatePatchForObject(original, modified, patchDocument, string.Empty);
+
+            return patchDocument;
+        }
+
+        /// <summary>
+        /// Recursive method to populate the Patch document for 2 objects.
+        /// </summary>
+        /// <param name="originalObject">The original version of the object.</param>
+        /// <param name="modifiedObject">The modified version of the object.</param>
+        /// <param name="patchDocument">The PATCH document to populate.</param>
+        /// <param name="path">The path to use for PATCH document paths.</param>
+        /// <typeparam name="T">The object type.</typeparam>
+        private static void PopulatePatchForObject<T>(T? originalObject, T? modifiedObject, JsonPatchDocument patchDocument, string path)
+            where T : class
+        {
+            // Use whichever object is no null to get the properties on the object
+            // If both are null then we create an empty list
+            var nonNullObject = originalObject ?? modifiedObject;
+            var properties = nonNullObject is null ? new List<PropertyInfo>() : nonNullObject.GetType()
+                .GetProperties(BindingFlags.FlattenHierarchy | BindingFlags.Instance | BindingFlags.Public).ToList();
+
+            // Now we loop through each property and try and create the PATCH document entry for it
+            foreach (var property in properties)
+            {
+                // Get the values of each property
+                var originalValue = originalObject != null ? property.GetValue(originalObject) : null;
+                var modifiedValue = modifiedObject != null ? property.GetValue(modifiedObject) : null;
+
+                if (originalValue is null && modifiedValue is not null)
+                {
+                    // Simplest case where the original was null and new isn't, we just add the new value
+                    patchDocument.Add($"{path}/{property.Name}", modifiedValue);
+                }
+                else if (originalValue is not null && modifiedValue is null)
+                {
+                    // Reverse of the above, we just remove
+                    patchDocument.Remove($"{path}/{property.Name}");
```

Comment:
🔨 Not sure if I understand this properly - but if we're patching a different property, couldn't this property have been passed as null within the patch object - and we end up overwriting it - essentially deleting the original property value? Or would an unintentional null property be caught before this?

Reply:
Oh, I think I get it now - this is removing the line from the patch document, not the value of the property from the object, right?

Reply:
Yes exactly this is saying, originally some value on the object was not null and now it is, so we add a "Remove" operation to the patch document. It doesn't do anything to the original object